gateway-service/src/main/java/org/zowe/apiml/gateway/services/ServicesInfoService.java

```diff
@@ -0,0 +1,269 @@
+                .description(instanceInfo.getMetadata().get(SERVICE_DESCRIPTION))
+                .homePageUrl(getGatewayUrl(instanceInfo.getHomePageUrl(), instanceInfo.getAppName().toLowerCase(), ServiceType.UI, routes))
+                .build();
+    }
+
+    private List<Authentication> getAuthentication(List<InstanceInfo> appInstances) {
+        return appInstances.stream()
+                .map(instanceInfo -> {
+                    Authentication authentication = eurekaMetadataParser.parseAuthentication(instanceInfo.getMetadata());
+                    return authentication.isEmpty() ? null : authentication;
+                })
+                .filter(Objects::nonNull)
+                .distinct()
+                .collect(Collectors.toList());
+    }
+
+    private Map<String, ServiceInfo.Instances> getInstances(List<InstanceInfo> appInstances) {
+        return appInstances.stream()
+                .collect(Collectors.toMap(
+                        InstanceInfo::getInstanceId,
+                        instanceInfo -> ServiceInfo.Instances.builder()
+                                .status(instanceInfo.getStatus())
+                                .hostname(instanceInfo.getHostName())
+                                .ipAddr(instanceInfo.getIPAddr())
+                                .protocol(getProtocol(instanceInfo))
+                                .port(getPort(instanceInfo))
+                                .homePageUrl(instanceInfo.getHomePageUrl())
+                                .healthCheckUrl(getHealthCheckUrl(instanceInfo))
+                                .statusPageUrl(instanceInfo.getStatusPageUrl())
+                                .customMetadata(getCustomMetadata(instanceInfo.getMetadata()))
+                                .build()
+                ));
+    }
+
+    private List<ServiceInfo.ApiInfoExtended> filterByIdAndMajorVersion(List<ServiceInfo.ApiInfoExtended> completeList) {
+        if (completeList.isEmpty()) return Collections.emptyList();
+
+        completeList.sort(Comparator
+                .comparing(ServiceInfo.ApiInfoExtended::getApiId, Comparator.nullsFirst(Comparator.naturalOrder()))
+                .thenComparing(ServiceInfo.ApiInfoExtended::getVersion, Comparator.comparing(this::getVersion))
+        );
+
+        List<ServiceInfo.ApiInfoExtended> result = new ArrayList<>(completeList.size());
+        ServiceInfo.ApiInfoExtended lastApiInfo = null;
+        for (ServiceInfo.ApiInfoExtended apiInfo : completeList) {
+            if (lastApiInfo == null ||
+                    !apiInfo.getApiId().equals(lastApiInfo.getApiId()) ||
+                    (apiInfo.getApiId().equals(lastApiInfo.getApiId()) &&
+                            getVersion(apiInfo.getVersion()).getMajorVersion() > getVersion(lastApiInfo.getVersion()).getMajorVersion())
+            ) {
+                result.add(apiInfo);
+                lastApiInfo = apiInfo;
+            }
+        }
+
+        return result;
```

Comment:
What dozou think about this code, is it the same logic? If you recognize it as better, use it, otherwise nevermind
```suggestion
	Map<String, Set<Version>> versions = new HashMap<>();
	for (ServiceInfo.ApiInfoExtended apiInfo : completeList) {
		Set<Version> v= versions.get(apiInfo.getApiId());
		if (v == null) {
			v = new HashSet<>();
		}
		v(getVersion(apiInfo.getVersion());
	}
	List<Version> all = new LinkedList<>();
	for (Set<Version> v : versions) {
		all.addAll(v);
	}
	return all;
```

Reply:
Or what about @plavjanik 's suggestion to use groupingBy https://www.baeldung.com/java-groupingby-collector?